File: app/src/main/java/com/github/typ0520/codee/route/InnerInvocationHandler.java
package com.github.typ0520.codee.route;

import com.github.typ0520.codee.network.route.RouteRetrofit;

import java.lang.reflect.InvocationHandler;
import java.lang.reflect.Method;
import retrofit2.PlatformDelegate;

/**
 * Created by tong on 2017/12/21.
 */
public class InnerInvocationHandler implements InvocationHandler {
    private final RouteRetrofit routeRetrofit;

    public InnerInvocationHandler(RouteRetrofit routeRetrofit) {
        this.routeRetrofit = routeRetrofit;
    }

    @Override
    public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
        if (method.getDeclaringClass() == Object.class) {
            return method.invoke(this, args);
        }
        if (PlatformDelegate.isDefaultMethod(method)) {
            return PlatformDelegate.invokeDefaultMethod(method, method.getDeclaringClass(), proxy, args);
        }
        Object realService = routeRetrofit.getRealService(proxy,method,args);
        return method.invoke(realService,args);
    }
}

File: app/src/main/kotlin/com/github/typ0520/codee/network/route/InnerInvocationHandler.kt
package com.github.typ0520.codee.network.route

import retrofit2.PlatformDelegate
import java.lang.reflect.InvocationHandler
import java.lang.reflect.Method

class InnerInvocationHandler(val routeRetrofit: RouteRetrofit): InvocationHandler {
    override fun invoke(proxy: Any, method: Method, vararg args: Any): Any? {
        if (method.declaringClass == Object::class.java) {
            return method.invoke(proxy, args)
        }
        if (PlatformDelegate.isDefaultMethod(method)) {
            return PlatformDelegate.invokeDefaultMethod(method, method.declaringClass, proxy, args)
        }
        val realService = routeRetrofit.getRealService(proxy, method, args)
        return method.invoke(realService, *args)
    }
}